File: ADT/djikstra's algo/djikstra's_alog.cpp

// Tc = O((V + E) log V) (vertices + edges * logV for push and pop from pq)
// Sc = O(V+E) (distance array + pq)
class Solution {
public:
    vector<int> dijkstra(int V, vector<vector<int>> &edges, int src) {
        // Build adjacency list
        vector<vector<pair<int,int>>> adj(V);
        for (auto &e : edges) {
            int u = e[0], v = e[1], w = e[2];
            adj[u].push_back({v, w});
            // adj[v].push_back({u, w}); // uncomment if graph is undirected
        }

        // Min-heap (distance, node)
        priority_queue<pair<int,int>,
                       vector<pair<int,int>>,
                       greater<pair<int,int>>> pq;

        vector<int> result(V, INT_MAX);
        result[src] = 0;
        pq.push({0, src});

        while (!pq.empty()) {
            auto [d, node] = pq.top();
            pq.pop();

            if (d > result[node]) continue;

            for (auto &p : adj[node]) {
                int adjNode = p.first;
                int wt = p.second;

                if (d + wt < result[adjNode]) {
                    result[adjNode] = d + wt;
                    pq.push({result[adjNode], adjNode});
                }
            }
        }

        return result;
    }
};


// by ordered set(ascending order store element)

/* 
📌 Why use set instead of priority_queue?
✅ Pros

We can erase outdated entries (priority_queue cannot do erase).

Cleaner logic (only one entry per node remains).

No need to check for outdated distances frequently.

❌ Cons

Slower constant factors than priority queue.

set operations are heavier (tree rotations & pointer chasing).
*/

class Solution {
public:
    vector<int> dijkstra(int V, vector<vector<int>> &edges, int src) {

        // Build adjacency list
        vector<vector<pair<int,int>>> adj(V);
        for (auto &e : edges) {
            int u = e[0], v = e[1], w = e[2];
            adj[u].push_back({v, w});
            // adj[v].push_back({u, w}); // Uncomment for undirected graph
        }

        // Set stores pairs {distance, node}
        // Automatically sorted by distance
        set<pair<int,int>> st;

        vector<int> dist(V, INT_MAX);
        dist[src] = 0;

        // Insert source node with distance 0
        st.insert({0, src});

        while (!st.empty()) {
            // Get the node with the smallest distance
            auto it = st.begin(); 
            int d = it->first;
            int node = it->second;

            st.erase(it);  // remove this entry

            // Traverse neighbors
            for (auto &p : adj[node]) {
                int adjNode = p.first;
                int wt = p.second;

                // Relaxation step
                if (d + wt < dist[adjNode]) {

                    // If the adj node already has an entry in set, remove it
                    if (dist[adjNode] != INT_MAX) {
                        st.erase({dist[adjNode], adjNode});
                    }

                    // Update the distance
                    dist[adjNode] = d + wt;

                    // Insert updated pair
                    st.insert({dist[adjNode], adjNode});
                }
            }
        }

        return dist;
    }
};

File: ADT/djikstra's algo/djikstra's_alog.py
# FlipKart, Amazon

import heapq

class Solution:
    def dijkstra(self, V, edges, src):
        """
        Dijkstra's algorithm using adjacency list and min-heap.
        V: number of vertices
        edges: list of [u, v, w]
        src: source node
        """

        # Build adjacency list
        adj = [[] for _ in range(V)]
        for u, v, w in edges:
            adj[u].append((v, w))
            # adj[v].append((u, w))  # uncomment for an undirected graph

        # Min heap (distance, node)
        pq = []
        heapq.heappush(pq, (0, src))

        # Initialize distance array
        dist = [float('inf')] * V
        dist[src] = 0

        while pq:
            d, node = heapq.heappop(pq)

            # Skip outdated entries
            if d > dist[node]:
                continue

            # Explore neighbors
            for adjNode, wt in adj[node]:
                if d + wt < dist[adjNode]:
                    dist[adjNode] = d + wt
                    heapq.heappush(pq, (dist[adjNode], adjNode))

        return dist


# using set
# for can erase preview entry of set but cannot do in pq

from sortedcontainers import SortedSet

class Solution:
    def dijkstra(self, V, edges, src):
        """
        Dijkstra using a SortedSet (balanced BST).
        """

        # Build adjacency list
        adj = [[] for _ in range(V)]
        for u, v, w in edges:
            adj[u].append((v, w))
            # adj[v].append((u, w))  # for undirected graph

        # (distance, node)
        st = SortedSet()
        dist = [float("inf")] * V
        dist[src] = 0

        # Insert source node
        st.add((0, src))

        while st:
            # Get smallest element
            d, node = st[0]
            st.discard((d, node))  # remove it

            # Process neighbors
            for adjNode, wt in adj[node]:
                if d + wt < dist[adjNode]:

                    # Remove old entry if exists
                    if dist[adjNode] != float("inf"):
                        st.discard((dist[adjNode], adjNode))

                    # Update distance
                    dist[adjNode] = d + wt
                    st.add((dist[adjNode], adjNode))

        return dist
